Metrics: add config duration recorder

martinkennelly opened the pull request:
Provides a mechanism to record config duration within OVN-K
for kubernete kinds from when event handler fires for a given
resource until processing is complete.
Exposes the duration as a histogram.

Optionally, provides a way to record OVN duration.

Currently added for the following k8 kinds:
* pod
* networkpolicy
* service


Signed-off-by: Martin Kennelly <mkennell@redhat.com>

martinkennelly commented:
/retest
Odd, I cannot repro locally.

martinkennelly commented:
/retest

martinkennelly commented:
/retest

martinkennelly commented:
This patch is missing a mechanism to detect hung OVN-Controllers. Will add later if this approach is acceptable.

squeed commented:
You are right, there is a race condition between transactions, but there's a way we can get around this. Just use a compare-and-swap to 

We need to use two variables: the newest nv_cfg and the newest *committed* nb_cfg. Then the code looks like this

```go
func transactOVS() {
    txid :=  GetNextNbCfgValue()
    do_ovs_commit()
    ReportNbCfg(txid)
}

var nextNB uint64
commitedLock sync.Mutex{}
var committedNB uint64

func GetNextNbCfgValue() {
     	return int(atomic.AddInt64(&nextNB, 1))
}

func ReportNbCfg(txid uint64) {
    commitedLock.Lock()
    defer committedLock.Unlock()
    if committedNB < txid {
        committedNB = txid
    }
}
```

Then, the metric we report is committedNB. There is a bit of a "race condtion" there where we might pull a newer "downstream" nb_cfg metric, but metrics have sloppy timing semantics to begin with, so this is not a big deal.

squeed commented on go-controller/pkg/metrics/recorder/recorder.go:
```diff
@@ -0,0 +1,489 @@
+func (cpr *Recorder) Run(sbClient, nbClient libovsdbclient.Client, kube kube.Interface, stop <-chan struct{}) {
+	cpr.Pod.run(sbClient, stop)
+	if config.Monitoring.EnableConfigDurationMonitor {
+		// OVN-Kube master process sends txs to OVN NB DB and if config duration monitor is enabled, it also adds
```
Killer comment. Awesome. Thanks for writing this.

martinkennelly replied:
All credit to Dumitru Ceara for explaining it to me.

bmeng commented on go-controller/pkg/metrics/recorder/recorder.go:
```diff
@@ -0,0 +1,494 @@
+
+func (cm *Config) run(nbClient libovsdbclient.Client, kube kube.Interface, stop <-chan struct{}) {
+	// Override command line to disable this functionality if cluster node size is large to protect SB DB
+	if isNodeCountGreaterThan(kube, configDurationNodeLimit) {
```
I am reconsidering this from SLO pov.
Compare with the `cpr.Pod.run(sbClient, stop)`, how much more load will be generated by this?

bmeng replied:
And since we have the flag `ovnkube_config_duration_monitor_enable` to enable/disable this.
Should we just leave the flag as a single switch?

martinkennelly replied:
> I am reconsidering this from SLO pov.
Compare with the cpr.Pod.run(sbClient, stop), how much more load will be generated by this?

From OVNs POV, very little. It must send out events to ovnkube-master for port_binding table.

>And since we have the flag ovnkube_config_duration_monitor_enable to enable/disable this.
Should we just leave the flag as a single switch?

Id rather have separate switches but since I didnt add a switch originally, I am not introducing one in this PR.

npinaeva commented on go-controller/pkg/metrics/recorder/recorder.go:
```diff
@@ -0,0 +1,494 @@
+// cItem is the data structure for config duration monitor queue
+type cItem struct {
+	op        operation
+	timestamp uint64
```
may useful to add a comment for units here (milliseconds)?

martinkennelly commented on go-controller/pkg/metrics/recorder/recorder.go:
```diff
@@ -0,0 +1,489 @@
+}
+
+// GetNextNbCfgValue returns the next, every increasing nb_cfg value
+func (cm *Config) GetNextNbCfgValue() int {
```
I've asked folks and it seems to be ok for cluster node size less than 50, but there is no guarantee there so its probably best not to add the operation to every Tx for now. I will rework. Thanks Casey.

martinkennelly commented:
/retest

martinkennelly commented:
Forced push to remove a binary that I accidentally committed ..

npinaeva commented on go-controller/pkg/metrics/master.go:
```diff
@@ -477,99 +507,239 @@ const (
 }
 
-// item is the data structure for ControlPlaneRecorders queue
+type podEvent struct {
```
would be nice to add some comments for podEvent vs configEvent

martinkennelly replied:
Resolved because I split the functionality out into separate structs.

martinkennelly commented:
WIP to split out the 2 separate functions in CPR.

npinaeva commented on go-controller/pkg/metrics/master.go:
```diff
@@ -477,99 +507,239 @@ const (
 }
 
 type ControlPlaneRecorder struct {
+	// pod monitor
```
what do you think about splitting ControlPlaneRecorder into something like PodRecorder and ConfigDurationRecorder?
and you can probably use a simple channel instead of a workqueue, since this recorder doesn't use all the extra features of a workqueue

martinkennelly replied:
Totally - will do.

npinaeva commented on go-controller/pkg/metrics/master.go:
```diff
@@ -640,45 +1021,29 @@ func (cpr *ControlPlaneRecorder) processNextItem() bool {
+		cpr.podRecords[i.p.uid] = &record{timestamp: i.p.timestamp, timestampType: firstSeen}
 	case cleanPod:
-		cpr.cleanPod(i.podUID)
+		delete(cpr.podRecords, i.p.uid)
```
Maybe we can also add deletion of a PodRecord after it's used the last time (which is after `metricPortBindingUpLatency.Observe`)

npinaeva commented on go-controller/pkg/metrics/master.go:
```diff
@@ -477,99 +507,239 @@ const (
+			newRow := new.(*nbdb.NBGlobal)
+
+			if oldRow.HvCfg != newRow.HvCfg && oldRow.HvCfgTimestamp != newRow.HvCfgTimestamp && newRow.HvCfgTimestamp > 0 {
+				if !cpr.queueFull() {
```
I think we should not ignore events from the databases (both nb and sb) even if the queue is full, because it can break the chain of events, e.g. missing LSP event will lead to all the rest of pod events being ignored because of the wrong `TimestampType`, and for config duration it can lead to the wrong duration being recorded if `hv_cfg` update was missed for the previous record
But we still can use it for adding-record events like AddPod or RecordConfigDuration

martinkennelly replied:
Makes sense. Will do.

npinaeva commented on go-controller/pkg/metrics/master.go:
```diff
@@ -477,99 +507,239 @@ const (
-			cpr.queue.Add(item{op: updatePortBinding, old: old, new: new, t: time.Now()})
-		},
-		DeleteFunc: func(table string, model model.Model) {
+			if !cpr.queueFull() {
```
since for `PortBinding` we only care about changes for `Chassis` and `Up` columns, will it make sense to check here if one of these values has changed here before putting new item to the queue?

martinkennelly replied:
done

martinkennelly commented:
Adding test cases and documentation now.

martinkennelly commented:
This PR should leverage this PRs new metrics "section" https://github.com/ovn-org/ovn-kubernetes/pull/2923 once it is merged.

martinkennelly commented:
/retest

flavio-fernandes commented on go-controller/pkg/metrics/master.go:
```diff
@@ -685,6 +713,374 @@ func getPodUIDFromPortBinding(row *sbdb.PortBinding) kapimtypes.UID {
+	// We increment the nb_cfg integer value in the NB_Global table.
+	// ovn-northd notices the nb_cfg change and copies the nb_cfg value to SB_Global table field nb_cfg along with any
+	// other configuration that is changed in OVN Northbound database.
+	// All ovn-controllers detects nb_cfg value change and generates a 'barrier' on the openflow connection to the
```
NIT: All ovn-controllers`detect` nb_cfg value...

npinaeva commented on go-controller/pkg/metrics/master.go:
```diff
@@ -681,6 +713,374 @@ func getPodUIDFromPortBinding(row *sbdb.PortBinding) kapimtypes.UID {
+				select {
+				case cdr.endRecordCh <- endRecordEvent{hvCfg: newRow.HvCfg, hvCfgTimestamp: newRow.HvCfgTimestamp}:
+				default:
+					klog.Warning(chFullMsg)
```
Should we delete corresponding `startRecord` here, because otherwise the next `hv_cfg` update will evaluate the wrong config duration for it?

martinkennelly replied:
Nice spot. Will add update.

npinaeva commented on go-controller/pkg/metrics/master_test.go:
```diff
@@ -0,0 +1,360 @@
+
+var _ = ginkgo.Describe("Config duration recorder operations", func() {
+	ginkgo.Context("recordConfigDuration()", func() {
+		ginkgo.It("should record a result for kind metric", func() {
```
will it make sense to merge this and the next test and verify correct recoring with a single test?

martinkennelly commented:
I note for reviewers that substantial changes are being made to handler logic here: https://github.com/ovn-org/ovn-kubernetes/pull/2909
The author and I discussed that this logic will not hamper my efforts, and I can rebase if that merges before me.

squeed commented on go-controller/pkg/metrics/master.go:
```diff
@@ -681,6 +713,374 @@ func getPodUIDFromPortBinding(row *sbdb.PortBinding) kapimtypes.UID {
+	cdr.reportKind[PodKindName] = prometheus.NewHistogram(prometheus.HistogramOpts{
+		Namespace: MetricOvnkubeNamespace,
+		Subsystem: MetricOvnkubeSubsystemMaster,
+		Name:      "network_programming_pod_duration_seconds",
```
I realize this is somewhat of a large change this late in the game, but a thought occurs to me: it would be a bit more idiomatic for this to be a single metric, with the Kind indicated by a label. Then we can aggregate on general latency, or drill down in to specific Kinds as necessary.

That would also suggest making the PodRecorder generic across object types.

martinkennelly replied:
yeah, that would be nicer. Its not too late as I am going to wait for #2909 as it seems more important than this. Thank you.

martinkennelly replied:
@squeed I just realised I would have to give up using buckets if I want to use labels.
I believe buckets do not support labelling. 
I can use a gauge or count vector.

squeed replied:
@martinkennelly https://pkg.go.dev/github.com/prometheus/client_golang/prometheus#NewHistogramVec should do the job...

martinkennelly replied:
I am an idiot.. Thanks casey

squeed commented:
I'm uncomfortable with automatically disabling this metric if the number of nodes goes over 100 -- this means that end-users could see their metrics randomly stop working for no obvious reason.

I think it would be better for there to be a sampling rate that is lower based on the number of nodes. In other words, by default, record latency for every 1/K requests, tuning K downwards as the number of nodes climbs. This avoids a sudden and abrupt loss of visibility.

martinkennelly commented:
>I'm uncomfortable with automatically disabling this metric if the number of nodes goes over 100 -- this means that end-users could see their metrics randomly stop working for no obvious reason.
I think it would be better for there to be a sampling rate that is lower based on the number of nodes. In other words, by default, record latency for every 1/K requests, tuning K downwards as the number of nodes climbs. This avoids a sudden and abrupt loss of visibility.

Yeah, its horrible. I did it to be uber cautions and protect large nodes. I will remove and only enable downstream when its tested with a large number of nodes.

martinkennelly commented on docs/metrics.md:
```diff
@@ -0,0 +1,17 @@
+|--|--|--|
+|ovnkube_master_network_programming_ovn_duration_seconds| Histogram  | The upper bound duration in seconds between a transaction sent to OVN from OVN-Kubernetes master and the application of this transaction to all nodes.
+|ovnkube_master_network_programming_pod_duration_seconds| Histogram  | The upper bound duration in seconds to configure a pod from when the pod handler function is executed and its configuration is applied to all nodes.
+|ovnkube_master_network_programming_service_duration_seconds| Histogram  | The upper bounded duration in seconds to configure a service from when OVN-Kubernetes master starts processing it following popping it from a work queue to when its configuration is applied to all nodes.
```
Its too much for now but we can add later.

martinkennelly commented:
Murali Krishnasamy will do a scale testing to see if resource p99s are degraded by this change.

martinkennelly commented:
/cc @trozet

ricky-rav commented on go-controller/pkg/ovn/obj_retry.go:
```diff
@@ -416,23 +417,76 @@ func (oc *Controller) getResourceFromInformerCache(objType reflect.Type, key str
+		klog.V(5).Infof("Recording add event on network policy")
+		np := obj.(*knet.NetworkPolicy)
+		metrics.GetConfigDurationRecorder().Start("networkpolicy", np.Namespace, np.Name)
+	case factory.ServiceType:
```
So we didn't apply the retry logic to services, because their controller already implements level-driven logic.

Is there a particular reason why you included this case block here and in the following functions (`recordUpdateEvent`, `recordDeleteEvent`, `recordSuccessEvent`)? This case block will never be executed because `WatchResource` is not applied to services. When you see the keyword services here it's actually `factory.PeerServiceType`, which corresponds to the services impacted by a network policy and we watch them with `WatchResource`.

martinkennelly replied:
Mistake. I will remove.

ricky-rav commented on go-controller/pkg/ovn/obj_retry.go:
```diff
@@ -742,6 +800,13 @@ func (oc *Controller) iterateRetryResources(r *retryObjs, updateAll bool) {
+			initObj = entry.newObj
+		} else if entry.oldObj != nil {
+			initObj = entry.oldObj
+		}
```
OK, I understand now why you recently asked me if both newObj and oldObj can be `nil`. If you get an update on an object and you already have a retry entry for this object marked for deletion (i.e. its oldObj != nil), then deletion might fail, in which case you will end up with a retry entry with oldObj!= nil (it's the old instance you were trying to delete) and `newObj!=nil` (it's the new instance from the update event). The same goes for an add event on an object with an existing retry entry marked for deletion, if deletion fails.

Anyway, here I see that the whole purpose of `initObj` is to pass it to `recordSuccessEvent`, from which you extract namespace and object name, which are not allowed by the API server to vary across an update, that is, in retry terms,  between `oldObj` and `newObj`. So in the end this if/else block is OK as it is! :)

martinkennelly commented:
All credit to Murali Krishnasamy, here are the P99 for pod latency:
* 5.08s
* 4.7s

So far so good. Doesn't look like any degradation. I think I can increase the sample rate.

martinkennelly commented:
All credit to Murali Krishnasamy, here are the P99 for pod latency:
* 5.08s
* 4.7s

So far so good. Doesn't look like any degradation. I think I can increase the sample rate.

martinkennelly commented:
@dcbw Any time or interest for a review?

martinkennelly commented:
Force pushed in order to fix conflicts with new master branch changes.

martinkennelly commented:
Things I am looking for input on:
* Small nit but bucket sizes for histograms maybe overly large - I copied kube-proxy histogram bucket size for network programming but due to having a label, there will potentially be 5-10 series eventually. I think I should reduce the bucket sizes to compensate. 
* I think "k" should be exposed as an option and not hard coded.

trozet commented on go-controller/pkg/config/config.go:
```diff
@@ -976,6 +979,11 @@ var MetricsFlags = []cli.Flag{
 		Destination: &cliConfig.Metrics.NodeServerCert,
 	},
+	&cli.BoolFlag{
+		Name:        "enable-config-duration",
```
would it be better to name this "metrics-enable-config-duration"?

trozet commented on go-controller/pkg/metrics/master.go:
```diff
@@ -698,6 +728,520 @@ func getPodUIDFromPortBinding(row *sbdb.PortBinding) kapimtypes.UID {
+	deleteOVNMeasurementChSize = 50
+	processChSize              = 1000
+	nbGlobalTable              = "NB_Global"
+	// we currently do not handle overflow, so we stop serving when close to overflow
```
LOL yeah @dceara told me about this. We need to get that fixed in OVN. Maybe add a FIXME here.

martinkennelly replied:
done

trozet commented on go-controller/pkg/metrics/master.go:
```diff
@@ -698,6 +728,520 @@ func getPodUIDFromPortBinding(row *sbdb.PortBinding) kapimtypes.UID {
+	}
+	ops, err := nbClient.Where(nbGlobal).Mutate(nbGlobal, model.Mutation{
+		Field:   &nbGlobal.NbCfg,
+		Mutator: ovsdb.MutateOperationAdd,
```
clever

trozet commented on go-controller/pkg/metrics/master.go:
```diff
@@ -698,6 +728,520 @@ func getPodUIDFromPortBinding(row *sbdb.PortBinding) kapimtypes.UID {
+		cr.measurementsMu.Lock()
+		m, ok = cr.measurements[kindNamespaceName]
+		if !ok {
+			klog.V(5).Infof("Config duration recorder: expected a measurement entry. Call Start before AddOVN"+
```
should this be an error? When is this OK?

martinkennelly replied:
Yep. It should be an error and it should never occur.

trozet commented on go-controller/pkg/metrics/master.go:
```diff
@@ -698,6 +728,520 @@ func getPodUIDFromPortBinding(row *sbdb.PortBinding) kapimtypes.UID {
+	}
+	ovnStartTimestamp := time.Now()
+
+	return ops, func() {
```
why does this function need to return a callback, is this to not block txns from executing and then the delay is after the txn call?

martinkennelly replied:
We call back when a successful tx was completed, and this initiates the OVN recoding. We don't want to start a measurement in OVN when tx did not succeed.

trozet replied:
ok that makes sense thanks

trozet reviewed (commented):
its hard for me to review every line of this PR but in general it looks good. Just some comments/questions.

npinaeva commented on go-controller/pkg/metrics/master.go:
```diff
@@ -698,6 +728,520 @@ func getPodUIDFromPortBinding(row *sbdb.PortBinding) kapimtypes.UID {
+	complete bool
+	// nb_cfg value that started the measurement
+	nbCfg int
+}
```
Great comments for all the structures!

npinaeva commented on go-controller/pkg/metrics/master.go:
```diff
@@ -698,6 +728,520 @@ func getPodUIDFromPortBinding(row *sbdb.PortBinding) kapimtypes.UID {
+
+// allowedToMeasure determines if we are allowed to measure or not. To avoid the cost of synchronisation by using locks,
+// we use probability.
+func (cr *ConfigDurationRecorder) allowedToMeasure(kindNamespaceName string) bool {
```
add comment about `allowedToMeasure` requiring to return the same value for the same `kindNamespaceName`

npinaeva commented on go-controller/pkg/metrics/master.go:
```diff
@@ -246,6 +250,32 @@ var metricPortBindingUpLatency = prometheus.NewHistogram(prometheus.HistogramOpt
+	Namespace: MetricOvnkubeNamespace,
+	Subsystem: MetricOvnkubeSubsystemMaster,
+	Name:      "network_programming_duration_seconds",
+	Help:      "The duration to apply network control plane configuration",
```
Help: "The duration to apply network configuration for a kind (e.g. pod, service, networkpolicy). Configuration includes add, update and delete events combined in a single histogram."
^ the best I can come up with

npinaeva commented on go-controller/pkg/metrics/master.go:
```diff
@@ -698,6 +728,520 @@ func getPodUIDFromPortBinding(row *sbdb.PortBinding) kapimtypes.UID {
+	// time when Add is executed
+	startTimestamp time.Time
+	// if true, startTimestamp is valid
+	start bool
```
can't find how this is used?

martinkennelly replied:
Removed.

npinaeva commented on go-controller/pkg/metrics/master.go:
```diff
@@ -693,6 +728,520 @@ func getPodUIDFromPortBinding(row *sbdb.PortBinding) kapimtypes.UID {
+// measurement. This must be called between Start and End. Not every call to Start will result in a measurement
+// and the rate of measurements depends on the number of nodes and function Run arg k.
+// Only one measurement for a kind/namespace/name is allowed until the current measurement is Ended (via End) and
+// processed. This is guaranteed by workqueues (even with multiple workers) and event handlers.
```
What workqueues do you mean? ConfigDurationRecorder doesn't use them right?

martinkennelly replied:
Workqueues by anything that uses this functionality i.e. services.

npinaeva commented on go-controller/pkg/metrics/master.go:
```diff
@@ -693,6 +728,520 @@ func getPodUIDFromPortBinding(row *sbdb.PortBinding) kapimtypes.UID {
+				continue
+			}
+			ovnKDelta = m.endTimestamp.Sub(m.startTimestamp).Seconds()
+			metricNetworkProgramming.With(prometheus.Labels{"kind": m.kind}).Observe(ovnKDelta)
```
metricNetworkProgramming then can include ovn-k only measurements, combined with e2e measurements, why would we do that?
Also, why have you decied to switch from only calling AddOVN with startTimestamp to more complicated Start->AddOVN->End chain of calls?

martinkennelly replied:
>metricNetworkProgramming then can include ovn-k only measurements, combined with e2e measurements, why would we do that?

I want to capture the full e2e duration for each kind. This metric has lower value if I separate it out into just the CMS and OVN as folks care about the e2e duration more than just CMS or OVN alone.
If folks want to look at the OVN duration, there is a dedicated metric for that independent of Kind.

>Also, why have you decied to switch from only calling AddOVN with startTimestamp to more complicated Start->AddOVN->End chain of calls?

I didn't like passing the timestamp through numerous layers of functions. Also, I couldnt start measuring before workqueues.

martinkennelly commented:
Forced pushed to address comments:
* Updated binary flag name
* Updates metrics.md to give a better description that nadia produced
* Removed "start" bool from measurement struct - unused
* Added Fixme and bz number for overflow bug in ovn
* Adjusted measurement output via klog in processMeasurements func to make it clearer (i hope).

trozet reviewed (approved):
/lgtm

Martin is on PTO so any necessary fix ups or additions we can do as another PR. This is a great start. Thanks @martinkennelly
